spacy/kb.pyx

```diff
@@ -0,0 +1,147 @@
+# cython: profile=True
+# coding: utf8
+from spacy.errors import user_warning
+
+
+cdef class Candidate:
+
+    def __init__(self, KnowledgeBase kb, entity_id_hash, alias_hash, prior_prob):
+        self.kb = kb
+        self.entity_id_hash = entity_id_hash
+        self.alias_hash = alias_hash
+        self.prior_prob = prior_prob
+
+    property entity_id:
+        """RETURNS (uint64): hash of the entity's KB ID"""
+        def __get__(self):
+            return self.entity_id_hash
+
+    property entity_id_:
+        """RETURNS (unicode): ID of this entity in the KB"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.entity_id]
+
+    property entity_name:
+        """RETURNS (uint64): hash of the entity's KB name"""
+        def __get__(self):
+            entry_index = <int64_t>self.kb._entry_index.get(self.entity_id)
+            return self.kb._entries[entry_index].entity_name_hash
+
+    property entity_name_:
+        """RETURNS (unicode): name of this entity in the KB"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.entity_name]
+
+    property alias:
+        """RETURNS (uint64): hash of the alias"""
+        def __get__(self):
+            return self.alias_hash
+
+    property alias_:
+        """RETURNS (unicode): ID of the original alias"""
+        def __get__(self):
+            return self.kb.vocab.strings[self.alias]
+
+    property prior_prob:
+        def __get__(self):
+            return self.prior_prob
+
+
+cdef class KnowledgeBase:
+
+    def __init__(self, Vocab vocab):
+        self.vocab = vocab
+        self._entry_index = PreshMap()
+        self._alias_index = PreshMap()
+        self.mem = Pool()
+        self._create_empty_vectors()
+
+    def __len__(self):
+        return self.get_size_entities()
+
+    def get_size_entities(self):
+        return self._entries.size() - 1  # not counting dummy element on index 0
+
+    def get_size_aliases(self):
+        return self._aliases_table.size() - 1 # not counting dummy element on index 0
+
+    def add_entity(self, unicode entity_id, unicode entity_name=None, float prob=0.5, vectors=None, features=None):
+        """
+        Add an entity to the KB.
+        Return the hash of the entity ID at the end
+        """
+        if not entity_name:
+            entity_name = entity_id
+        cdef hash_t id_hash = self.vocab.strings.add(entity_id)
+        cdef hash_t name_hash = self.vocab.strings.add(entity_name)
+
+        # Return if this entity was added before
+        if id_hash in self._entry_index:
+            user_warning("Entity " + entity_id + " already exists in the KB")
```

Comment:
Small code style edit:

```suggestion
            user_warning("Entity {} already exists in the KB".format(entity_id))
```

Soo looking forward to the day when we can switch the entire code base over to f-strings. (I think it shouldn't be a problem in Cython already, but we want things to be consistent so regular string formatting it is for now.)

**Edit:** Ah, just realised that this will kinda solve itself once the messages are moved to `spacy.errors`. Nevermind.